server/src/main/java/org/elasticsearch/index/query/IntervalsSourceProvider.java

```diff
@@ -0,0 +1,463 @@
+                && Objects.equals(ordered, match.ordered);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(text, maxWidth, ordered);
+        }
+
+        @Override
+        public String getWriteableName() {
+            return NAME;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeString(text);
+            out.writeInt(maxWidth);
+            out.writeBoolean(ordered);
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.startObject(NAME);
+            builder.field("text", text);
+            builder.field("max_width", maxWidth);
+            builder.field("ordered", ordered);
+            return builder.endObject().endObject();
+        }
+
+        private static final ConstructingObjectParser<Match, Void> PARSER = new ConstructingObjectParser<>(NAME,
+            args -> {
+                String text = (String) args[0];
+                int max_width = (args[1] == null ? Integer.MAX_VALUE : (Integer) args[1]);
+                boolean ordered = (args[2] == null ? false : (Boolean) args[2]);
+                return new Match(text, max_width, ordered);
+            });
+        static {
+            PARSER.declareString(constructorArg(), new ParseField("text"));
+            PARSER.declareInt(optionalConstructorArg(), new ParseField("max_width"));
+            PARSER.declareBoolean(optionalConstructorArg(), new ParseField("ordered"));
+        }
+
+        public static Match fromXContent(XContentParser parser) throws IOException {
+            return PARSER.apply(parser, null);
+        }
+    }
+
+    public static class Combine extends IntervalsSourceProvider {
+
+        public static final String NAME = "combine";
+
+        protected enum Type {
+            ORDERED {
+                @Override
+                IntervalsSource getSource(List<IntervalsSource> subSources) {
+                    return Intervals.ordered(subSources.toArray(new IntervalsSource[0]));
+                }
+            }, UNORDERED {
+                @Override
+                IntervalsSource getSource(List<IntervalsSource> subSources) {
+                    return Intervals.unordered(subSources.toArray(new IntervalsSource[0]));
+                }
+            }, OR {
+                @Override
+                IntervalsSource getSource(List<IntervalsSource> subSources) {
+                    return Intervals.or(subSources.toArray(new IntervalsSource[0]));
+                }
+            };
+
+            abstract IntervalsSource getSource(List<IntervalsSource> subSources);
+        }
+
+        private final List<IntervalsSourceProvider> subSources;
+        private final Type type;
+        private final int maxWidth;
```

Comment:
Should we extract this option to have a dedicated source for `max_width` ? Or could it be handled by a scripted intervals source ?

Reply:
I was originally going to do this as a separate source, but it only really makes sense on ordered/unordered combinations, so I think it's useful to have it here?

Reply:
I'd prefer a dedicated source but I can live with it. I agree that it makes sense on ordered/unordered sources but it can also be useful for others so we could also have a dedicated source **and** an option in the `match` source ?